tap_mailchimp/streams.py

```diff
@@ -0,0 +1,538 @@
+    path = '/members'
+    streams_to_sync = ['lists']
+    data_key = 'members'
+    bookmark_path = ['lists', '', stream_name, 'datetime']
+    bookmark_query_field = 'since_last_changed'
+    replication_keys = ['last_changed']
+
+
+class ListSegments(FullTable):
+    """Class for 'list_segments' stream"""
+    stream_name = 'list_segments'
+    key_properties = ['id']
+    path = '/segments'
+    streams_to_sync = ['lists']
+    data_key = 'segments'
+    child = ['list_segment_members']
+
+
+class ListSegmentMembers(FullTable):
+    """Class for 'list_segment_members' stream"""
+    stream_name = 'list_segment_members'
+    key_properties = ['id']
+    path = '/members'
+    streams_to_sync = ['lists', 'list_segments']
+    data_key = 'members'
+
+
+class Campaigns(FullTable):
+    """Class for 'campaigns' stream"""
+    stream_name = 'campaigns'
+    key_properties = ['id']
+    path = '/campaigns'
+    params = {
+        'status': 'sent',
+        'sort_field': 'send_time',
+        'sort_dir': 'ASC'
+    }
+    child = ['reports_email_activity', 'unsubscribes']
+    report_streams = ['reports_email_activity']
+    data_key = stream_name
+
+
+class Unsubscribes(FullTable):
+    """Class for 'unsubscribes' stream"""
+    stream_name = 'unsubscribes'
+    key_properties = ['campaign_id', 'email_id']
+    path = '/reports/{}/unsubscribed'
+    streams_to_sync = ['campaigns']
+    data_key = stream_name
+
+
+class ReportEmailActivity(Incremental):
+    """Class for 'reports_email_activity' stream"""
+    stream_name = 'reports_email_activity'
+    extra_fields = ['emails.activity']
+    key_properties = ['campaign_id', 'action', 'email_id', 'timestamp']
+    streams_to_sync = ['campaigns']
+    path = '/reports/{}/email-activity'
+    data_key = 'emails'
+    replication_keys = ['timestamp']
+
+    def write_activity_batch_bookmark(self, batch_id):
+        """Write batch id bookmark"""
+        self.write_bookmark(['reports_email_activity_last_run_id'], batch_id)
+
+    def write_email_activity_chunk_bookmark(self, current_bookmark, current_index, sorted_campaigns):
+        """Write chunk bookmark for email activities"""
+        # Bookmark the next chunk because the current chunk will be saved in batch_id
+        # Index is relative to current bookmark
+        next_chunk = current_bookmark + current_index + 1
+        if next_chunk * EMAIL_ACTIVITY_BATCH_SIZE < len(sorted_campaigns):
+            self.write_bookmark(['reports_email_activity_next_chunk'], next_chunk)
+        else:
+            self.write_bookmark(['reports_email_activity_next_chunk'], 0)
+
+    def get_batch_info(self, batch_id):
+        """Function to get batch status"""
+        try:
+            return self.client.get('/batches/{}'.format(batch_id), endpoint='get_batch_info')
+        except HTTPError as e:
+            if e.response.status_code == 404:
+                raise BatchExpiredError('Batch {} expired'.format(batch_id))
+            raise e
```

Comment:
Unit test for the exception part is missing.

Reply:
Added unittest for raising BatchExpiredError.